illumos-utils/src/zfs.rs

```diff
@@ -611,32 +835,37 @@ impl Zfs {
             id,
             additional_options,
         }: DatasetEnsureArgs,
-    ) -> Result<(), EnsureDatasetError> {
+    ) -> Result<(), EnsureDatasetErrorRaw> {
         let (exists, mounted) = Self::dataset_exists(name, &mountpoint)?;
 
         let props = build_zfs_set_key_value_pairs(size_details, id);
         if exists {
             // If the dataset already exists: Update properties which might
             // have changed, and ensure it has been mounted if it needs
             // to be mounted.
-            Self::set_values(name, props.as_slice()).map_err(|err| {
-                EnsureDatasetError {
-                    name: name.to_string(),
-                    err: err.err.into(),
-                }
-            })?;
-
-            let want_to_mount = !zoned
-                && matches!(can_mount, CanMount::On)
-                && matches!(mountpoint, Mountpoint::Path(_));
+            Self::set_values(name, props.as_slice())
+                .map_err(|err| EnsureDatasetErrorRaw::from(err.err))?;
 
-            if !mounted && want_to_mount {
-                Self::mount_dataset(name)?;
+            if !zoned && !mounted {
+                if let (CanMount::On, Mountpoint::Path(path)) =
```

Comment:
What other kinds of `Mountpoint`s are there? (And what would it mean if we had `CanMount::On, MountPoint::SomethingElse(_)`?)